lib/Controller/ImageController.php

```diff
@@ -0,0 +1,210 @@
+				$downloadResult = $this->imageService->insertImageLink($documentId, $link, $userId);
+			}
+			return new DataResponse($downloadResult);
+		} catch (Exception $e) {
+			$this->logger->error('Link insertion error', ['exception' => $e]);
+			return new DataResponse(['error' => 'Link insertion error'], Http::STATUS_BAD_REQUEST);
+		}
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @PublicPage
+	 *
+	 * @param int $documentId
+	 * @param int $sessionId
+	 * @param string $sessionToken
+	 * @param string|null $shareToken
+	 * @return DataResponse
+	 */
+	public function uploadImage(int $documentId, int $sessionId, string $sessionToken, ?string $shareToken = null): DataResponse {
+		if (!$this->sessionService->isValidSession($documentId, $sessionId, $sessionToken)) {
+			return new DataResponse([], Http::STATUS_INTERNAL_SERVER_ERROR);
+		}
+
+		try {
+			$file = $this->request->getUploadedFile('image');
+			if ($file !== null && isset($file['tmp_name'], $file['name'], $file['type'])) {
+				if (!in_array($file['type'], self::IMAGE_MIME_TYPES, true)) {
+					return new DataResponse(['error' => 'Image type not supported'], Http::STATUS_BAD_REQUEST);
+				}
+				$newFileContent = file_get_contents($file['tmp_name']);
+				$newFileName = $file['name'];
+				if ($shareToken) {
+					$uploadResult = $this->imageService->uploadImagePublic($documentId, $newFileName, $newFileContent, $shareToken);
+				} else {
+					$session = $this->sessionService->getSession($documentId, $sessionId, $sessionToken);
+					$userId = $session->getUserId();
+					$uploadResult = $this->imageService->uploadImage($documentId, $newFileName, $newFileContent, $userId);
+				}
+				return new DataResponse($uploadResult);
+			}
+			return new DataResponse(['error' => 'No uploaded file'], Http::STATUS_BAD_REQUEST);
+		} catch (Exception $e) {
+			$this->logger->error('Upload error', ['exception' => $e]);
+			return new DataResponse(['error' => 'Upload error'], Http::STATUS_BAD_REQUEST);
+		}
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 * @PublicPage
+	 *
+	 * Serve the images in the editor
+	 * @param int $documentId
+	 * @param int $sessionId
+	 * @param string $sessionToken
+	 * @param string $imageFileName
+	 * @param string|null $shareToken
+	 * @return DataDisplayResponse
+	 */
+	public function getImage(int $documentId, int $sessionId, string $sessionToken, string $imageFileName, ?string $shareToken = null): DataDisplayResponse {
+		if (!$this->sessionService->isValidSession($documentId, $sessionId, $sessionToken)) {
+			return new DataDisplayResponse('', Http::STATUS_NOT_FOUND);
+		}
+
+		try {
+			if ($shareToken) {
+				$imageFile = $this->imageService->getImagePublic($documentId, $imageFileName, $shareToken);
+			} else {
+				$imageFile = $this->imageService->getImage($documentId, $imageFileName, $this->userId);
+			}
+			return $imageFile !== null
+				? new DataDisplayResponse($imageFile->getContent(), Http::STATUS_OK, ['Content-Type' => $imageFile->getMimeType()])
```

Comment:
I'm not fuilly sure if necessary, but to be on the safe side this should use https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/lib/public/Files/IMimeTypeDetector.php#L68 to avoid any mimetype that would allow potential XSS

Reply:
:+1: Done in 7628714403f41706f5dd9465c7a4b48898a6ecbc